### DISCUSSION TITLE
Can I use variadic arguments of different types and can I use V without linking system libs (including vlib)?

### DISCUSSION THREAD

**Discussion by rempas** [vlang/v#13110] [General]
Hello everyone, new to V! I want to make 2 questions.

1. Can I use arguments of different types for variadic functions? I tried something like this using `any`:
```
fn test_fn(args ...any) {
  for arg in args {
    print("Args: $arg")
  }
}

fn main() {
  test_fn("Hello", "John", 10, true)
}
```
 But I'm getting the following error: error: no known default format for type `any`. So I suppose this will not work as like the documentation says, this is like `void*` so it needs to dereference it or something.

2. This is simple, but can I link my program without `vlib` and without any other system libraries (like when passing "-nostdlib" in TCC/GCC/Clang etc.)?

**Comment by JalonSolov** [comment]
From `v help build-c` output:
```
   -freestanding
      Build the executable without dependency on libc.
      Supported only on `linux` targets currently.
```
The only way to link without `vlib` is to not include any calls to anything in `vlib`... which is impossible, since you will always use the `builtin` module at a minimum.

**Reply by rempas** [reply]
Thanks, well I was thinking about creating my own library in assembly so that's why this question

**Comment by Wajinn** [comment]
Maybe try using generics with the variadic, but the different parameters must be of the same type.

```
fn test_gen_vari<T>(args ...T) {
	for arg in args  {
		print("Args: $arg ")
	}
}

fn main() {
	test_gen_vari('Hello', 'John', 'Cat')
	test_gen_vari(8, 10, 12, 16, 3)
}
```

**Reply by rempas** [reply]
Yeah, I thought about that but I was wondering if someone knew if I can have them be different types from one another. Thanks a lot for your time

**Reply by Wajinn** [reply]
> Yeah, I thought about that but I was wondering if someone knew if I can have them be different types from one another. Thanks a lot for your time

Apologies for not answering this previously, as after posting the generic variadic, a better answer had occurred to me (using sum types) but it had slipped my mind to post it. 

```
type Mixer = int | string | bool   // put the different types or structs that you want

fn main() {
	println(mix_type("Hello", "John", 10, true))  
}

// function is accepting arguments of different types

fn mix_type(list ...Mixer) ([]string, []int, []bool) {
	mut nums := []int{}
	mut strs := []string{}
	mut bools := []bool{}
	
	for value in list {
		if value is int {
			nums << value
		}
		if value is string {
			strs << value
		}
		if value is bool	{
			bools << value
		}
	}
	return strs, nums, bools
}
```

**Reply by rempas** [reply]
Don't worry. We are all here in our free time and nobody is obliged to help anyone. So I'm grateful that you found the time to do so now ;)

When it comes to the reply, sum types are great but like I said, to my other post about "printf", I would like to see a way to pass types of different types without having to create a sum type manually. Maybe in the future!
